src/backend/distributed/planner/recursive_planning.c

```diff
@@ -455,6 +466,97 @@ ShouldRecursivelyPlanSubquery(Query *subquery, RecursivePlanningContext *context
 }
 
 
+/*
+ * ShouldRecursivelyPlanSetOperation determines whether the leaf queries of a
+ * set operations tree need to be recursively planned in order to support the
+ * (sub)query as a whole.
+ */
+static bool
+ShouldRecursivelyPlanSetOperation(Query *query, RecursivePlanningContext *context)
+{
+	PlannerRestrictionContext *filteredRestrictionContext = NULL;
+
+	SetOperationStmt *setOperations = (SetOperationStmt *) query->setOperations;
+	if (setOperations == NULL)
+	{
+		return false;
+	}
+
+	if (context->level == 0)
+	{
+		/*
+		 * We cannot push down top-level set operation. Recursively plan the
+		 * leaf nodes such that it becomes a router query.
+		 */
+		return true;
+	}
+
+	if (setOperations->op != SETOP_UNION)
+	{
+		/*
+		 * We can only push down UNION operaionts, plan other set operations
+		 * recursively.
+		 */
+		return true;
+	}
+
+	if (DeferErrorIfUnsupportedUnionQuery(query) != NULL)
+	{
+		/*
+		 * If at least one leaf query in the union is recurring, then all
+		 * leaf nodes need to be recurring.
+		 */
+		return true;
+	}
+
+	filteredRestrictionContext =
+		FilterPlannerRestrictionForQuery(context->plannerRestrictionContext, query);
```

Comment:
It looks like the only problematic function for local tables is  `FilterPlannerRestrictionForQuery()->  ReferenceRelationCount() -> PartitionMethod()`. In fact, we should have skipped that for local tables when we allowed local tables in subqueries.

Reply:
Changed `AddRteRelationToAttributeEquivalenceClass`  to skip local relations.